text/0000-offset_of.md

```diff
@@ -0,0 +1,576 @@
+   `mem::size_of` or `mem::align_of`, but plausibly greater as it provides even
+   more information.
+
+    That said, this API allows querying information which (if needed at compile
+    time) would otherwise be hard-coded, so in some cases it may reduce the risk
+    of a compatibility hazard.
+
+2. Similarly, this reduces the amount of dynamism that a Rust implementation
+   could use for `repr(Rust)` types.
+
+    For example, it forbids a Rust implementation from varying field offsets of
+    `repr(Rust)` types between executions of the same compiled program (for
+    example, by way of interpretation or code modification), unless it also
+    performs modifications to adjust the result of `offset_of!` (and recompute
+    the values of derived constants, and regenerate relevant code, ...).
+
+3. This is a feature most code won't need to use, and it may be confusing to
+   users unfamiliar with low level programming.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The general rationale is that it should remove the need to hardcode, hand-roll,
+or pull in a third-party crate in order to compute field offsets. This hopefully
+should remove as many barriers
+
+That said, there are several alternatives to this, some of which were even
+considered:
+
+1. Do nothing, and tell users to use the [`memoffset`][memoffset] crate, or to
+   hard-code constant offsets.
+
+    This was not chosen as this operation seems fundamental enough to provided
+    by the standard library, especially given how often it is incorrectly
+    implemented in the wild.
+
+2. Add `offset_of!`, but disallow use on `#[repr(Rust)]` types.
+
+    This would make `core::mem::offset_of!` have less functionality than the
+    implementation from `memoffset`, or the implementation they could implement
+    if they computed it manually.
+
+    Needing the offset of fields on `#[repr(Rust)]` is not as common, but still
+    useful in code which needs to describe precise details of type layout to
+    some other system, including GPU APIs which accept configurable vertex
+    formats or binary serialization formats that contain descriptions of the
+    field offsets for the record types they contain, etc.
+
+    It is also useful for implementing field projection as a library feature, as
+    in cases like [`field-offset`][fieldoffset].
+
+3. Require that all fields of `$Container` be visible at the invocation site,
+   rather than just requiring that `$field` is.
+
+    As above, this would make `core::mem::offset_of!` worse than the version
+    they'd have written themselves and/or an off-the-shelf implementation.
+
+4. Add `offset_of!`, but disallow use during constant evaluation.
+
+    This would mean that users which need const access to `offset_of!` must
+    continue to hardcode the field offsets as constants, which is undesirable,
+    error-prone, and can cause compatibility hazards.
```

Comment:
As an aside, [my example](https://github.com/rust-lang/rfcs/pull/3308#issuecomment-1242688535) is also a reason to allow using the macro in values that are subject to _static promotion_. This is currently a stronger requirement than being usable in constant evaluation — see e.g. the `#[rustc_promotable]` annotation on functions like `size_of` that makes them usable in static promotion values.

Reply:
The future for static promotion is `const { ... }` blocks. There are no plans to make any more operations subject to implicit static promotion; if anything, we'll continue trying to take back things that have been implicitly promoted without enough thought or care (https://github.com/rust-lang/rust/issues/80619).